File: caesar.f90
! Bob Nisco
! Theory of Programming Languages
! Spring 2013

program Caesar
	implicit none

	! Declare some variables
	character(45) :: str = "The quick brown fox jumped over the lazy dog"
	integer, parameter :: shftAmt = 4

	! Make str all upper case
	call toUpper(str)
	call encrypt(str, shftAmt)
	write(*, *) str
	call decrypt(str, shftAmt)
	write(*, *) str
	call solve(str, 26)
	contains

	subroutine encrypt(str, shftAmt)
		character(*), intent(inout) :: str
		integer, intent(in) :: shftAmt
		integer :: i

		do i = 1, len(str)
			if (iachar(str(i:i)) == 32) then
				str(i:i) = achar(32)
			else
				str(i:i) = achar(modulo(iachar(str(i:i)) - 65 + shftAmt, 26) + 65)
			endif
		end do

	end subroutine encrypt

	subroutine decrypt(str, shftAmt)
		character(*), intent(inout) :: str
		integer, intent(in) :: shftAmt
		integer newShftAmt

		! Decrypt is really encrypt inverse
		newShftAmt = shftAmt * -1
		call encrypt(str, newShftAmt)
	end subroutine decrypt

	subroutine solve(str, maxShftAmt)
		character(*), intent(inout) :: str
		integer, intent(in) :: maxShftAmt
		character(len=len(str)) :: strCopy
		integer i

		do i = 0, maxShftAmt
			strCopy = str
			call encrypt(strCopy, i)
			write(*, *) "Caesar ", i, ": ", strCopy
		end do
	end subroutine

	subroutine toUpper(str)
		implicit none
		character(*), intent(inout) :: str
		integer i
		integer j

		do i = 1, len(str)
			j = iachar(str(i:i))
			if (j>= iachar("a") .and. j<=iachar("z")) then
				str(i:i) = achar(iachar(str(i:i)) - 32)
			else
				str(i:i) = str(i:i)
			end if
		end do
	end subroutine toUpper

end program Caesar

File: Caesar.scala
/**
 * Caesar Cipher Scala Class and CaesarTest object for testing the Caesar class
 *
 * @author Bob Nisco
 * Theory of Programming Languages
 * Spring 2013
 */
class Caesar {

	// Caesar will have an attribute of str
	var str = ""

	/**
	 * @param shftAmt the amount of characters that str should be shifted
	 * @return an encryptd string based on the given shftAmt
	 */
	def encrypt(shftAmt: Int): String = {
		val newShftAmt = shftAmt % 26
		val charArray = this.str.toUpperCase().toCharArray()
		var ret = ""

		for (i<-charArray) {
			if (i.toInt == 32) {
				ret += " ";
			} else {
				var temp = (i.toInt - 65 + newShftAmt) % 26
				if (temp < 0) {
					temp += 26
				}
				ret += (temp + 65).toChar
			}
		}

		return ret
	}

	/**
	 * @param shftAmt the amount of characters that str should be shifted
	 * @return a decryptd string based on the given shftAmt
	 */
	def decrypt(shftAmt: Int): String = {
		return encrypt(shftAmt * - 1)
	}

	/**
	 * @param maxShftAmt the maximum amount of different combinations of the decrypt
	 *        function that should be iterated over.
	 * @return void, but it will print out the decrypt functions from 0 to maxShftAmt
	 */
	def solve(maxShftAmt: Int) {
		for (i <- 0 to maxShftAmt) {
			println("Caesar " + i + ": " + encrypt(i))
		}
	}
}

object CaesarTest {
	/**
	 * @param args an array of String arguments
	 */
	def main(args: Array[String]) {
		var caesar = new Caesar()
		caesar.str = "FATURE RAM IS OTILL I"
		caesar.str = caesar.encrypt(4)
		println(caesar.str)
		caesar.str = caesar.decrypt(4)
		println(caesar.str)
		println(caesar.solve(26))
	}
}
